### ISSUE TITLE
🐛 Bug Report: Memory leak while uploading in Firefox with multiple Pingvin-Share tabs open

### ISSUE THREAD

**Issue by axiomen** [stonith404/pingvin-share#340]
### 👟 Reproduction steps

I've noticed a memory leak with larger file uploads that only happens under these conditions:

1. Browser: Firefox.
2. More than one tab of Pingvin-Share open.
3. Start uploading a large file.
4. RAM starts rapidly filling up.

I tested under Edge and Chrome, but the issue seems to be present only in Firefox. #165 might be related.

### 👍 Expected behavior

The same behavior across all browsers.

### 👎 Actual Behavior

Memoy leak only in Firefox.

### 🌐 Browser

Firefox

**Comment by stonith404** [comment]
@axiomen I can't reproduce this anymore. While you did the testing of the "premature close" error, did you still notice this memory leak?

**Comment by axiomen** [comment]
Yes, I've just tested it. First test was an upload with 2.1GB and 2.5GB files, multiple tabs open. The second test was an upload with two 4.8GB files, only one tab, however the single tab was opened after anothe Pingvin tab was closed, and then the upload started (so I'm guessing the old tab was still in memory).

![memory_leak](https://github.com/stonith404/pingvin-share/assets/150123148/e8aa605a-25b6-41d4-bd6f-b7273b2d9b0b)

This consumes my entire RAM, so I'm guessing it would have gone even higher if I had more memory. It doesn't go down on its own, I have to restart the browser. If there's only one tab freshly opened tab in a browser session, the memory goes up, but never that high, and it self-regulates, and empties once the upload is done.

I cannot recreate this on Chrome, plus the memory footprint is much, much smaller (doesn't go above 500 MB even with large files, while in Firefox, it can grow 4GB+).

---

On another note, I did get a Premature error on the latest build as well, so might need to re-open #354 :/

**Comment by axiomen** [comment]
Yes, I've just tested it. First test was an upload with 2.1GB and 2.5GB files, multiple tabs open. The second test was an upload with two 4.8GB files, only one tab, however the single tab was opened after anothe Pingvin tab was closed, and then the upload started (so I'm guessing the old tab was still in memory).

![memory_leak](https://github.com/stonith404/pingvin-share/assets/150123148/e8aa605a-25b6-41d4-bd6f-b7273b2d9b0b)

This consumes my entire RAM, so I'm guessing it would have gone even higher if I had more memory. It doesn't go down on its own, I have to restart the browser (on some occassions it will go down on its own, not sure why). If there's only one tab freshly opened tab in a browser session, the memory goes up, but never that high, and it self-regulates, and empties once the upload is done.

I cannot recreate this on Chrome, plus the memory footprint is much, much smaller (doesn't go above 500 MB even with large files, while in Firefox, it can grow 4GB+).

On further testing, there might be some weird interaction with addons that causes the leak. However, the large memory footprint remains an issue in Firefox even without the leak.

---

On another note, I did get a Premature error on the latest build as well, so might need to re-open #354 :/

**Comment by stonith404** [comment]
Sorry for my late response. Thanks for checking this again. You bring bad luck as you are reporting the most difficult bugs 😉. This memory leak and the premature error really give me a headache. I will look into it and try to fix it when I find some free time.

**Comment by axiomen** [comment]
Ahah, I have to stress test it for production 😅

I did notice the same thing happening on my other PC, on a Linux distro. I have 64GB of RAM on that one, and while uploading a file, half of the RAM was eaten by Pingvin. The file was smaller than the amount of RAM taken. It went back down shortly after the upload was finished. The only difference between having multiple Pingvin instances open and just a single one is, as I've reported here, most of the time the RAM doesn't clear out once the file has been uploaded. In general, there is some funky interaction between Pingvin and Firefox.
